src/ServiceControl.LearningTransport/LearningTransportNativeQueueLengthProvider.cs

```diff
@@ -0,0 +1,88 @@
+﻿namespace ServiceControl.Transports.LearningTransport
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Linq;
+    using System.Threading.Tasks;
+    using Monitoring;
+    using Monitoring.Infrastructure;
+    using Monitoring.QueueLength;
+
+    public class LearningTransportNativeQueueLengthProvider : INativeQueueLengthProvider
+    {
+        string rootFolder;
+        ConcurrentDictionary<string, LearningTransportQueueLengthTracker> trackers = new ConcurrentDictionary<string, LearningTransportQueueLengthTracker>(StringComparer.InvariantCultureIgnoreCase);
+
+        public Dictionary<EndpointInstanceId, int> SnapshotValues()
+        {
+            return (
+                from t in trackers.Values
+                from c in t.Update()
+                group c by c.Key
+                into g
+                select new KeyValuePair<EndpointInstanceId, int>(g.Key, g.Sum(x => x.Value))
+            ).ToDictionary(x => x.Key, x => x.Value);
+        }
+
+        public void TrackQueue(EndpointInstanceId endpointInstanceId, string queueName, string nodeId)
+        {
+            trackers.AddOrUpdate(queueName,
+                q =>
+                {
+                    var tracker = new LearningTransportQueueLengthTracker(rootFolder, q);
+                    tracker.Add(endpointInstanceId);
+                    return tracker;
+                },
+                (q, existing) =>
+                {
+                    existing.Add(endpointInstanceId);
```

Comment:
Is there a possibility that another thread will obtain the tracker and will try to do something else with it at the same time? In other words, is the tracker itself thread-safe and should it be?